datafusion/src/datasource/mod.rs

```diff
@@ -36,3 +47,231 @@ pub(crate) enum Source<R = Box<dyn std::io::Read + Send + Sync + 'static>> {
     /// Read data from a reader
     Reader(std::sync::Mutex<Option<R>>),
 }
+
+#[derive(Debug, Clone)]
+/// A single file that should be read, along with its schema, statistics
```

Comment:
Do I understand it that the idea here is to allow a DataFusion reader to read a set of files in a generic way (rather than having separate implementations for csv, parquet, etc)?

Reply:
Yes, that's the intention here.
- `PartitionedFile` -> Single file (for the moment) or part of a file (later, part of the row groups or rows), and we may even extend this to include partition value and partition schema (see below) to support partitioned tables:
`/path/to/table/root/p_date=20210813/p_hour=1200/xxxxx.parquet`
- `FilePartition` -> The basic unit for parallel processing, each task is responsible for processing one `FilePartition` which is composed of several `PartitionFile`s.